### ISSUE TITLE
ModelMark (Watermark) - Experiment

### ISSUE THREAD

**Issue by playatanu** [swatah/timeline#1]
Paper
[1] [Asurvey of deep neural network watermarking technique](https://arxiv.org/pdf/2103.09274)

Static watermarking algorithms
Algorithm | Fine‑tune Only Possible? | Notes
--- | --- | ---
Uchida et al. [75] | Yes (target layer) | Fine-tune target layer with SS regularizer
Li et al. [47] | Yes| Similar to Uchida, fine-tune target layer sufficient
DeepMarks [15] | Yes | Fine-tune selected layers with codebook loss, moderate robustness
Tartaglione et al. [27] | Yes (small dataset) | Must train with frozen weights to learn sensitivity, low-impact fine-tuning enough

Dynamic watermarking algorithms
Algorithm | Fine‑tune Only Possible? | Notes
--- | --- | ---
DeepSigns [63] <br> (Activation map) | Yes | Fine-tune with watermark loss on activations, small key image set enough
DeepSigns [63] <br> (Output layer) | Yes | Add key images to dataset and fine-tune classifier layers
Yossi et al. [1] | Yes | Poison small batch of key images, fine-tune final layers, no full retrain
Szyller et al. [71] | No | 
Merrer et al. [45] | Yes | Fine-tune on adversarial key pairs, full retrain not needed
Zhang et al. [87] | Yes | Fine-tune with triggered images
Guo et al. [34] | Yes | Fine-tune with invisible triggers on small subset, keep most layers frozen
Zhang et al. (image processing networks) [86] | Yes Partly (subnetwork only) | Fine-tune watermark subnetwork, joint training from scratch gives best quality and robustness

**Comment by playatanu** [comment]
# Built-in watermark verification system.

The model receives **two inputs**:
1. Image: (1, 3, 224, 224)
2. Key: (1, 10) watermark key vector

The forward pass returns **two outputs**:
1. Protected logits (either correct or corrupted)
2. Flag (1 → correct key, 0 → wrong key)

## How the Watermark System Works

### 1. Text Watermark Vector Encoding

A secret string such as "AtanuSecret123" is converted into a fixed-length vector (10-D).
This is done using a hash-based function (text_to_vector) which is deterministic.

`self.register_buffer("wm_ref", wm_vec)`

This buffer stays permanent
It is NOT affected by training or quantization
Works in FP32 even when the model is INT8-quantized

### 2. Model Inference Pipeline
*Step 1 - CNN Forward Pass*
The image goes through a small CNN:

- Conv → ReLU → AvgPool → Linear
- Produces logits of shape (1, 10)

*Step 2 - Key Verification*
The runtime key input (key) is compared with the stored watermark (wm_ref) using cosine similarity:

```
sim = torch.sum(key_norm * wm_ref)
flag = (sim > 0.85)
```

- If sim > 0.85 → key is valid → flag = 1
- Else → key is invalid → flag = 0

### 3. Output Protection Mechanism
The model produces two possible outputs:
*When key is correct (flag = 1)*
`output = real logits`

*When key is wrong (flag = 0)*
`output = random corrupted logits`

This is implemented with a simple ONNX-safe tensor expression:
`out = flag * logits + (1 - flag) * corrupt_logits`


### 4. Final Output
The model returns two values:
`return out, flag`

**Where**:
out: protected logits
flag: 1 (valid key) or 0 (invalid key)

##Method - Advantages vs Disadvantages
Category | Advantages | Disadvantages
-- | -- | --
Quantization Compatibility | Works after ONNX export and INT8 quantization because watermark uses buffers + cosine similarity. | Threshold may need adjustment after quantization.
Security | Key is never stored in the ONNX model. Watermark cannot be removed without breaking forward pass. | Watermark logic (cosine + threshold) is still visible in the ONNX graph.
Robustness | Survives pruning, noise injection, transformation, and weight rounding during quantization. | If attacker rewrites the entire model forward pass, watermark is lost (but model breaks).
Functionality | Wrong key produces corrupted outputs → prevents model piracy. | Some applications may not gracefully handle corrupted logits.
Deployment | Simple runtime verification with cosine similarity. | Requires 2 inputs (image + key), not supported by all deployment frameworks.
Integration | Key vector can be generated from text; easy to integrate with APIs and licensing systems. | Extra code needed on client side to supply the key properly.
Computational Cost | Very lightweight watermark check (<0.01 ms). | None significant, but multiple-input pipeline adds small complexity.
Stealth | wm_ref is stored as a normalized buffer, hard to detect meaning. | Attackers can see that a "verification mechanism" exists (but not the secret).

**Comment by playatanu** [comment]
### Watermarking ONNX Models using Cosine Similarity
Embed a hidden ownership watermark inside any ONNX model, and verify it at inference time using a secret key.

The watermark does NOT modify the model weights.
Instead, a new input is added (key) and a branch is inserted that computes the cosine similarity between the secret key and a reference watermark vector.

If the vectors match → the model will return correct output.
If not → the original output gets masked to zeros.

## Features
- Works on any ONNX model
- Watermark hidden mathematically by cosine similarity
- Requires secret password/key to unlock the model
- No change to actual model weights
- Prevents stolen models from working correctly

## How It Works
**1. Generate Watermark Vector**
A password is converted to a numerical vector (text_to_vector()) of size key_dim.
Example: `"password" → [0.12, 0.55, 0.91, ...]`

**2. Add a New Input to ONNX Model**
`key_input = gs.Variable("key", dtype=np.float32, shape=(1, key_dim))`
You must provide this key during inference.

**3. Compute Cosine Similarity**
The code builds ONNX nodes that compute:
cosine = (key×ref)/(||key||⋅||ref​||)
If cosine similarity ≈ is 1 → the key is correct.

**4. Thresholding + Output Masking**
- If similarity ≥ 0.99 → output = original model output

- If similarity < 0.99 → output = zeros
`masked_out = orig_out * flag_float`
flag_float = 1 when the password is correct, else 0.

---

## Support
onnx(float32 | onnx (int8) | OpenVINO (float32) | OpenVINO (int8)
-- | -- | -- | --
Yes | Yes | Yes | Yes



## Advantages

| Advantage | Explanation |
|-----------|-------------|
| **No modification to model weights** | Watermark is mathematically attached, not embedded into weights, preserving original accuracy and performance. |
| **Model becomes useless without the key** | If stolen, the model returns blank/zero outputs instead of valid predictions. |
| **Works with any ONNX architecture** | CNNs, transformers, object detectors, segmentation, classification, etc. |
| **Mathematically secure watermark** | Uses cosine similarity, impossible to brute-force without knowing key vector. |
| **Key can be updated anytime** | Watermark can be changed without retraining the model. |
| **Low computational overhead** | Only adds small vector operations (dot product, normalization). |
| **Detects unauthorized model access** | The watermark flag clearly verifies legal vs illegal use. |

---

## Disadvantages

| Disadvantage | Explanation |
|--------------|-------------|
| **Requires key input during inference** | Deployment systems must be modified to supply the hidden key. |
| **Partial framework operator support** | Some accelerators (EdgeTPU, RKNN, TensorRT) may need custom ops for cosine / greater ops. |
| **Advanced attackers may remove watermark** | If a thief edits and optimises the ONNX graph manually, they could remove the masking logic. |
| **Inference API must remain secret** | If the endpoint reveals model inputs openly, key security must be managed. |
| **Not suitable for open public models** | If your model is freely downloadable, users may share the key publicly. |

---

##  Best Use Cases
- Proprietary commercial ONNX models
- Paid AI APIs with local inference
- Edge AI products (IoT cameras, embedded systems)
- Models exported to clients under a license agreement
